forch/port_state_manager.py

```diff
@@ -143,38 +143,54 @@ def handle_device_behavior(self, mac, device_behavior, static=False):
     def handle_device_placement(self, mac, device_placement, static=False):
         """Handle a learning or expired VLAN for a device"""
         if device_placement.connected:
-            # if device is learned
-            self._placement_to_mac[(device_placement.switch, device_placement.port)] = mac
-            self._process_device_placement(mac, device_placement, static=static)
+            return self._handle_learned_device(mac, device_placement, static)
 
-            if mac not in self._state_machines:
-                self._state_machines[mac] = PortStateMachine(
-                    mac, PortStateMachine.UNAUTHENTICATED, self._handle_unauthenticated_state,
-                    self._set_port_sequestered, self._set_port_operational,
-                    self._handle_infracted_state)
+        return self._handle_disconnected_device(device_placement)
 
-                device_behavior = (self._static_device_behaviors.get(mac) or
-                                   self._dynamic_device_behaviors.get(mac))
-                if device_behavior:
-                    static = mac in self._static_device_behaviors
-                    self.handle_device_behavior(mac, device_behavior, static=static)
+    def _handle_learned_device(self, mac, device_placement, static=False):
+        # if device is learned
+        old_mac = self._placement_to_mac.get((device_placement.switch, device_placement.port))
+        stale_mac = old_mac if old_mac and old_mac != mac else None
 
-            return True, None
+        if stale_mac:
+            switch = device_placement.switch
+            port = device_placement.port
+            self._logger.warning(
+                'Cleaning stale device placement: %s, %s, %s', old_mac, switch, port)
+            stale_placement = DevicePlacement(switch=switch, port=port, connected=False)
+            self._handle_disconnected_device(stale_placement)
+
+        self._placement_to_mac[(device_placement.switch, device_placement.port)] = mac
+        self._process_device_placement(mac, device_placement, static=static)
+
+        if mac not in self._state_machines:
+            self._state_machines[mac] = PortStateMachine(
+                mac, PortStateMachine.UNAUTHENTICATED, self._handle_unauthenticated_state,
+                self._set_port_sequestered, self._set_port_operational,
+                self._handle_infracted_state)
+
+            device_behavior = (self._static_device_behaviors.get(mac) or
+                               self._dynamic_device_behaviors.get(mac))
+            if device_behavior:
+                static = mac in self._static_device_behaviors
+                self.handle_device_behavior(mac, device_behavior, static=static)
+
+        return True, None, stale_mac
```

Comment:
note this is fine for now, but it's headed in a problematic direction -- in the long run, it makes things much harder to understand with complex tuple return values, esp. when the point of call just returns that again.  Since python (by default, how we use it) doesn't specify return types in the method signature, it means a lot of spaghetti knowledge in the code to figure things out.

Also, it's losing the parity with the function title ("handle") -- since it's doing a lot more than handle the event.

Anyway, fine for now, but next time you find yourself trying to overload a tuple return value, consider ways to restructure the code so that it doesn't need to work that way (there's a whole bunch of ways to accomplish that, just depending on the situation).